### PR TITLE
Added List.sort(comp) to List module

### PR DESCRIPTION
**Pull request by mathewmariani**
Added `sort(comp)` function to List using an in-place quicksort implementation.

### CODE REVIEW

**Comment by CohenArthur** [comment]
Your doc example shows it, but it would be nice to have some documentation on the expected behavior of `comp()`. For example, saying it must take two arguments and return a bool, etc etc

**Comment by mathewmariani** [comment]
> Your doc example shows it, but it would be nice to have some documentation on the expected behavior of `comp()`. For example, saying it must take two arguments and return a bool, etc etc

Thanks for pointing that out. I just updated the docs and added a default `comp` functions

**Comment by mhermier** [comment]
From my experience at using the language, I would recommend to change `sort()` to be a uni-line. Stub functions are only meant to provide default values or simply computed values, they should behave exactly like their parent and also return the same thing. While technically equivalent in your case, making it already there, express intent more, avoid possible future bugs and late fixing.

So as a rule of thumb; if you write a one liner, it is highly likely you want to return the computed value, so use the real one liner expression (that implicitly returns the computed value). I know, it is not really intuitive or does not scale producing changes, but the language is like this...

**Comment by mathewmariani** [comment]
> From my experience at using the language, I would recommend to change `sort()` to be a uni-line. Stub functions are only meant to provide default values or simply computed values, they should behave exactly like their parent and also return the same thing. While technically equivalent in your case, making it already there, express intent more, avoid possible future bugs and late fixing.
> 
> So as a rule of thumb; if you write a one liner, it is highly likely you want to return the computed value, so use the real one liner expression (that implicitly returns the computed value). I know, it is not really intuitive or does not scale producing changes, but the language is like this...

Just to clarify you mean a uni-line (stue function) like so `sort() { sort {|a, b| a < b} }` or is there a Wren style that im missing?

**Comment by mhermier** [comment]
Yes this one, it act as a stub with implicit returning.

File: test/core/list/sort.wren
Comment (bjorn) [inline test/core/list/sort.wren:2]: Does this test pass? I don't see `List.sort()` returning any value, so I would expect this to print `null`. Did you mean to `return this` at the end of `sort`?

Comment (mathewmariani) [reply test/core/list/sort.wren:2]: With this [issue](https://github.com/wren-lang/wren/issues/801) open the tests don't pass.

Comment (mhermier) [reply test/core/list/sort.wren:1]: @bjorn one liner body/method/function does implicitly return.

Comment (mathewmariani) [reply test/core/list/sort.wren:1]: @mhermier the tests were returning null

Comment (bjorn) [reply test/core/list/sort.wren:1]: @mhermier I know, but since the `sort(_)` function was returning `null`, so did the one-line version of `sort()`.

File: src/vm/wren_core.wren
Comment (bjorn) [inline src/vm/wren_core.wren:328]: I think it's rather weird for `sort()` to return `this` but not `sort(comparer)`. I'd suggest changing this back to the one-line version and putting the `return this` in `sort(comparer)`.

File: test/core/list/sort.wren
Comment (mathewmariani) [reply test/core/list/sort.wren:1]: So `sort(_)` should return `this` and the one-liner `sort()` should be reverted back to a stub function?

**Comment by mhermier** [comment]
Does the tests pass now ?

**Comment by mathewmariani** [comment]
> Does the tests pass now ?

Not this the [stack corruption](https://github.com/wren-lang/wren/issues/761) issue

**Comment by mhermier** [comment]
I think the test should be a bit more atomic to be more interesting.

**Comment by mathewmariani** [comment]
I rebased from `main` and can confirm the tests are passing

**Comment by bjorn** [comment]
@mathewmariani It seems you did a rebase followed by a merge. Please get rid of the merge for force-push just the rebased version of your branch.

**Comment by mathewmariani** [comment]
> @mathewmariani It seems you did a rebase followed by a merge. Please get rid of the merge for force-push just the rebased version of your branch.

I believe that's it

**Comment by mathewmariani** [comment]
Should I resolve all the comments, or leave that up to the commenters?

**Comment by bjorn** [comment]
@mathewmariani I see most of my comments marked as "Outdated" and I see no way to resolve them (neither on ones that haven't been marked outdated), so I guess that's on you.

**Comment by mathewmariani** [comment]
Anything else needed for this to be merged?

File: projects/xcode/wren_test.xcodeproj/project.xcworkspace/contents.xcworkspacedata
Comment (bjorn) [reply projects/xcode/wren_test.xcodeproj/project.xcworkspace/contents.xcworkspacedata:2]: You resolved this comment but the file is still being added by the pull request.

Comment (mathewmariani) [reply projects/xcode/wren_test.xcodeproj/project.xcworkspace/contents.xcworkspacedata]: @bjorn should be done now

**Comment by ChayimFriedman2** [comment]
Nice work. Three things, however:

 1. I think it's preferred to implement this in C. Wren prefers to implement the core in C (speed).
 2. You implemented a quick sort. Do we need a stable sort instead?
 3. Should we add a non in-place sort, too? (Maybe a method in `Sequence`). It can be just:
    ```wren
    class Sequence {
      // ...
      sort() { sort {|a, b| a < b } }
      sort(comparer) { toList.sort(comparer) }
    }
    ```

**Comment by avivbeeri** [comment]
I don't think it's semantically appropriate to have a Sequence.sort method, because a sequence could in theory be infinite in length (and they can be lazily evaluated), so a sorting operation would be impossible or intractable.

**Comment by ruby0x1** [comment]
Also a sort function in c runs into reentrancy (specifically, it's not possible when a user supplied compare function is involved)

**Comment by ChayimFriedman2** [comment]
@ruby0x1 I think the reentrancy is about to be solved, am I wrong?

@avivbeeri I know that, be that still holds in Python for example, and it does have a `sorted()` function that operates on iterables.

**Comment by mathewmariani** [comment]
@ChayimFriedman2 I implemented Quick Sort since it's the algorithm used in Lua's `table.sort`
@ruby0x1 is there anything left that I should be adding for this MR?

**Comment by ruby0x1** [comment]
I'll look closely soon, busy with some other Wren stuff locally.

File: src/vm/wren_core.wren
Comment (ruby0x1) [inline src/vm/wren_core.wren:337]: @mathewmariani this is using `<` despite comparer being provided. 
I would expect `if(comparer.call(a, b)) {` to be the condition instead?

File: doc/site/modules/core/list.markdown
Comment (ruby0x1) [inline doc/site/modules/core/list.markdown:106]: This doesn't explain the boolean result, or the nature of the compare function :) We'd ideally do both, I can adjust the documentation after merge though, just wanted to give feedback.

> The comparison function accepts two arguments, which are the two values to compare, and must return a boolean value to indicate if the first value is smaller than the second value. If the function returns true, the first argument will be before the second in the sorted results. 
> 
> A compare function like `{|value, other| true }` will always put `value` before `other`. The default compare function is `{|a, b| a < b }`.

File: src/vm/wren_core.wren
Comment (bjorn) [reply src/vm/wren_core.wren:337]: Actually `a` and `b` are the start and end indexes of the range that's being sorted. Since the range is inclusive, the comparison is checking whether the range has more than 1 element.

Maybe there could be more intuitive names for these variables.

File: src/vm/wren_core.wren
Comment (ruby0x1) [inline src/vm/wren_core.wren:326]: I suspect this might want to be `<=` for stable sort results, otherwise equal items will rearrange themselves on successive sorting, and won't be deterministic.

File: src/vm/wren_core.wren
Comment (ruby0x1) [reply src/vm/wren_core.wren:337]: Oh right, I misread there. I agree with better naming.

File: src/vm/wren_core.wren
Comment (ChayimFriedman2) [reply src/vm/wren_core.wren:326]: Quick sort isn't stable anyway.

Comment (ChayimFriedman2) [reply src/vm/wren_core.wren:326]: https://stackoverflow.com/questions/13498213/quicksort-algorithm-stability

Comment (avivbeeri) [reply src/vm/wren_core.wren:326]: It can be implemented in a manner that makes it stable, at the cost of memory consumption: https://www.geeksforgeeks.org/stable-quicksort/

As an alternative, we could use a merge sort, which has similar properties, plus stability. However, it would also have some memory consumption requirements.

Comment (ChayimFriedman2) [reply src/vm/wren_core.wren:326]: I know that, but unless we use a stable sort (which I suggested earlier), it's useless to discuss about `<` vs. `<=`.

**Comment by mhermier** [comment]
To help comprehension, could you replace 'a' and 'b' with something like
'i' and 'j' where iterators are expected.

**Comment by ChayimFriedman2** [comment]
I think `start` and `end` would be better.

**Comment by mathewmariani** [comment]
@ChayimFriedman2 @mhermier I'm leaning towards `i` and `j` since that's what you find in most books. I also like more compact code.

File: src/vm/wren_core.wren
Comment (mathewmariani) [reply src/vm/wren_core.wren:326]: Is there a specific reason we want a stable sort? My idea for using a non-stable quick-sort was that it's highly memory efficient which is perfect for, in my opinion, virtual memory environments like the WrenVM. It might have a worst-case scenario of running at O(n^2) but on average it should run O(n logn).

If a more efficient algorithm is needed it can be left to the user.

File: src/vm/wren_core.wren
Comment (mathewmariani) [reply src/vm/wren_core.wren:337]: I was thinking `i` and `j` since that's what most algorithms & data-structure books might use. But, `start` and `end` can be used as well.

File: src/vm/wren_core.wren
Comment (avivbeeri) [reply src/vm/wren_core.wren:326]: I think we can get away with it being unstable, so long as we document it clearly. From an embeddability point of view, the memory usage argument is quite compelling, in my opinion.

Comment (mathewmariani) [reply src/vm/wren_core.wren:326]: @avivbeeri Thank you! and yes, I will have to redo the documentation

**Comment by ruby0x1** [comment]
@mathewmariani code is for humans, not machines, legibility > compactness. especially for wren, which strives to help people learn and understand, making the code clearer is part of that goal. In my code I used `low` `high` but `start` and `end` is also clear enough. I would prefer not to use single letter variables here.

**Comment by ruby0x1** [comment]
Thanks @mathewmariani , I'll probably tweak some things after merging, but looks good from here.

**Review by ruby0x1 — APPROVED** [review]

**Pull request merged**
